# **Sentiment Analysis using NLP Pipeline & ML Models**

Build a complete end-to-end Sentiment Analysis system by applying NLP preprocessing, feature engineering, and multiple Machine Learning models. Compare their performance using evaluation metrics.


Text → Preprocessing → Vectorization → ML Model → Sentiment Prediction

## **STEP 1: Setup Environment**

In [8]:
!pip install nltk scikit-learn pandas numpy
import nltk

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

## **STEP 2: Load Dataset**

In [9]:
import pandas as pd

# upload dataset
df = pd.read_csv("IMDB_Dataset.csv", engine='python')
df.head()

,review,sentiment
0,One of the other reviewers has mentioned that ...,positive
1,A wonderful little production. <br /><br />The...,positive
2,I thought this was a wonderful way to spend ti...,positive
3,Basically there's a family where a little boy ...,negative
4,"Petter Mattei's ""Love in the Time of Money"" is...",positive


## **STEP 3: Data Understanding**

In [10]:
print(df.shape)
print(df.columns)
print(df['sentiment'].value_counts())

(50000, 2)
Index(['review', 'sentiment'], dtype='object')
sentiment
positive    25000
negative    25000
Name: count, dtype: int64


## **Step 4: NLP Preprocessing**

In [11]:
import nltk
import string
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer

nltk.download('punkt')
nltk.download('stopwords')

stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()

def preprocess_text(text):
    # 1. Lowercase
    text = text.lower()

    # 2. Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))

    # 3. Tokenization
    tokens = word_tokenize(text)

    # 4. Remove stopwords
    tokens = [word for word in tokens if word not in stop_words]

    # 5. Stemming
    tokens = [stemmer.stem(word) for word in tokens]

    return " ".join(tokens)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [12]:
df['clean_text'] = df['review'].apply(preprocess_text)

## **STEP 5: Feature Engineering**

#### **1. Bag of Words**

In [14]:
from sklearn.feature_extraction.text import CountVectorizer

bow = CountVectorizer()
X_bow = bow.fit_transform(df['clean_text'])

#### **2. TF-IDF**

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
X_tfidf = tfidf.fit_transform(df['clean_text'])

## **STEP 6: Train-Test Split**

In [16]:
from sklearn.model_selection import train_test_split

y = df['sentiment']

X_train, X_test, y_train, y_test = train_test_split(
    X_tfidf, y, test_size=0.2, random_state=42
)

## **STEP 7: Model Building**

#### **1. Logistic Regression**

In [17]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

**2. Naive Bayes**

In [18]:
from sklearn.naive_bayes import MultinomialNB

nb = MultinomialNB()
nb.fit(X_train, y_train)

y_pred_nb = nb.predict(X_test)

#### **3. Decision Tree**

In [19]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

## **STEP 8: Model Evaluation**

In [20]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def evaluate_model(y_test, y_pred, model_name):
    print(f"--- {model_name} ---")
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred, average='weighted'))
    print("Recall:", recall_score(y_test, y_pred, average='weighted'))
    print("F1 Score:", f1_score(y_test, y_pred, average='weighted'))
    print()

In [22]:
# Model Evaluation
evaluate_model(y_test, y_pred_lr, "Logistic Regression")
evaluate_model(y_test, y_pred_nb, "Naive Bayes")
evaluate_model(y_test, y_pred_dt, "Decision Tree")

--- Logistic Regression ---
Accuracy: 0.8923
Precision: 0.8926329041936555
Recall: 0.8923
F1 Score: 0.8922627388768944

--- Naive Bayes ---
Accuracy: 0.8641
Precision: 0.8644282778341221
Recall: 0.8641
F1 Score: 0.8640873675612795

--- Decision Tree ---
Accuracy: 0.7159
Precision: 0.71593268815723
Recall: 0.7159
F1 Score: 0.715904309856002



## **Step 9: Comparison**

## Model Comparison & Insights

### Observations

- **Logistic Regression** achieved the highest accuracy (**89.23%**) among all models.  
  It performs well because it can handle high-dimensional sparse data generated by TF-IDF and works efficiently for linear classification problems.

- **Naive Bayes** also performed well with an accuracy of **86.41%**.  
  It is fast and efficient, but it assumes that all features are independent, which is not always true in real-world text data.

- **Decision Tree** showed the lowest performance (**71.59% accuracy**).  
  This is because Decision Trees are not well-suited for sparse and high-dimensional data and tend to overfit.

---

### Key Insights

- Logistic Regression is the **best-performing model** for this dataset.  
- Naive Bayes is a good **baseline model** due to its speed and simplicity.  
- Decision Tree is **not ideal for NLP tasks** with vectorized text data.

---

### Trade-offs

- Logistic Regression → High accuracy but requires proper feature engineering  
- Naive Bayes → Fast but based on strong assumptions  
- Decision Tree → Easy to understand but prone to overfitting  

## **Step 10: Final Conclusion**

##  Final Conclusion

In this project, a complete NLP pipeline was implemented to perform sentiment analysis on textual data. The process included text preprocessing, feature engineering, model training, and evaluation.

- The **best model** was **Logistic Regression**, as it achieved the highest accuracy and balanced performance across all evaluation metrics.  
- The **best vectorization technique** used was **TF-IDF**, as it effectively captured important words and reduced the impact of common words.  

This project also highlights the importance of **text preprocessing**, as cleaning and transforming raw text significantly improves model performance.

Overall, the combination of **TF-IDF and Logistic Regression** proved to be the most effective approach for sentiment analysis in this task.

---

### Final Insight

A well-structured NLP pipeline is essential for converting raw text into meaningful features, which directly impacts the performance of machine learning models.